### PR TITLE
Fix aws elb hosted zone id add nlb ids

### LINKED ISSUE #7988: aws_elb_hosted_zone_id data source doesn't support Network Load Balancers
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

aws_elb_hosted_zone_id data source supports Application Load Balancers and Classic Load Balancers (which use the same hosted zone id), but not Network Load Balancers (their hosted zone ids are different, see: https://docs.aws.amazon.com/general/latest/gr/rande.html#elb_region).

As an example, the configuration defined below will return "Z32O12XQLNTSW2" as hosted ID zone, but it's impossible to retrieve network load balancer equivalent: "Z2IFOLAFXWLO4F".

### Potential Terraform Configuration

```hcl
data "aws_elb_hosted_zone_id" "main" {
    region = "eu-west-1"
}
```

### References

* https://docs.aws.amazon.com/general/latest/gr/rande.html#elb_region
* #1618

### PR DESCRIPTION
**Pull request by Jukie**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #7988

Changes proposed in this pull request:

* Add `elb_type` argument to support Network Load Balancers
* Update acceptance tests for new argument
* Update web docs

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSElbHostedZoneId_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSElbHostedZoneId_basic -timeout 120m
=== RUN   TestAccAWSElbHostedZoneId_basic
=== PAUSE TestAccAWSElbHostedZoneId_basic
=== CONT  TestAccAWSElbHostedZoneId_basic
--- PASS: TestAccAWSElbHostedZoneId_basic (11.49s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	11.612s

...
```

### CODE REVIEW

**Comment by Jukie** [comment]
Updated my branch to let `elb_type` be an optional argument which defaults to application/classic if not set.

**Comment by carinadigital** [comment]
@Jukie could you fix the conflicts ?

**Comment by Jukie** [comment]
Sure I'll fix them now, any ideas why us-gov-west-1's elb hosted zone id was removed? https://github.com/terraform-providers/terraform-provider-aws/commit/2954561946419cba3f973b9904bf2df959cf9d38#diff-e47acbc3ed755feb1ae98b9f5b6a7115L29

**Comment by Jukie** [comment]
Updated with correct NLB id's as well as upstream changes to drop us-gov-west-1 due to what appears to be a lack of documentation.

**Comment by Jukie** [comment]
Updated test run below:
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSElbHostedZoneId_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSElbHostedZoneId_basic -timeout 120m
=== RUN   TestAccAWSElbHostedZoneId_basic
=== PAUSE TestAccAWSElbHostedZoneId_basic
=== CONT  TestAccAWSElbHostedZoneId_basic
--- PASS: TestAccAWSElbHostedZoneId_basic (28.06s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	28.161s
```

**Comment by Jukie** [comment]
Should be ready now @carinadigital

File: aws/data_source_aws_elb_hosted_zone_id.go [hunk 1]
```diff
@@ -50,9 +65,18 @@ func dataSourceAwsElbHostedZoneIdRead(d *schema.ResourceData, meta interface{})
 	if v, ok := d.GetOk("region"); ok {
 		region = v.(string)
 	}
-
+	var i int
```
Comment (carinadigital) [inline aws/data_source_aws_elb_hosted_zone_id.go]: Maybe rename variable to something more descriptive.
lbTypeIndex ?

Comment (Jukie) [reply aws/data_source_aws_elb_hosted_zone_id.go]: Thanks for the suggestion, I'll add that now.

**Comment by Jukie** [comment]
Updated acctest after variable changes:
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSElbHostedZoneId_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSElbHostedZoneId_basic -timeout 120m
=== RUN   TestAccAWSElbHostedZoneId_basic
=== PAUSE TestAccAWSElbHostedZoneId_basic
=== CONT  TestAccAWSElbHostedZoneId_basic
--- PASS: TestAccAWSElbHostedZoneId_basic (27.70s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	27.759s
```

**Comment by Jukie** [comment]
Updated acctest:
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSElbHostedZoneId_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSElbHostedZoneId_basic -timeout 120m
=== RUN   TestAccAWSElbHostedZoneId_basic
=== PAUSE TestAccAWSElbHostedZoneId_basic
=== CONT  TestAccAWSElbHostedZoneId_basic
--- PASS: TestAccAWSElbHostedZoneId_basic (27.06s)
PASS
```

**Comment by bflad** [comment]
Hey @Jukie 👋 Thanks for working on this! It would be really nice to get the NLB hosted zone information somewhere available in the provider.

I'm personally _very_ torn on whether we should expand the existing data source (and potentially adding confusion with "ELB" in its naming) versus just adding a new data source (call it `aws_nlb_hosted_zone` or `aws_nlb_hosted_zone_id` although we've been shying away from `_id` naming since data sources may have other attributes in the future, poor `aws_subnet_ids` 😅). If we go the new data source route, we can also create a `aws_alb_hosted_zone` data source that points to the same underlying information as `aws_elb_hosted_zone_id`.

How do you feel about leaving the existing data source and just adding this as a new `aws_nlb_hosted_zone` data source?

**Comment by Jukie** [comment]
Well "ELB" _technically_ encompasses Classic, ALB, and NLB if I'm not mistaken but I also recognize the possible confusion because when people say ELB they're usually referring to the "Classic" type.
I'd be fine with splitting them. Would you also happen to know if there's any api call or managed endpoint that would return the list of id's at https://docs.aws.amazon.com/general/latest/gr/rande.html? Unless these are guaranteed to stay the same, we should probably be validating the values we're hardcoding in the data source as well.

**Comment by shmick** [comment]
I went looking for `aws_nlb_hosted_zone_id`and came across this PR. Is there any update on this?

**Comment by Jukie** [comment]
@bflad In my opinion splitting them will be a duplication of code so adding a type attribute with a default value of classic/alb to avoid breaking existing user's config seems like a good option but I'm happy to hear alternatives.

- The term ELB encompasses ALB, Classic, and NLB so it also seems like only splitting off NLB could create confusion.

- This data source has a single purpose of returning an elb hosted zone id so I'd also expect that if there's any feature/attribute expansion it would happen in the `aws_lb` or `aws_elb` data sources and if needed we'd make the determination about splitting the "main" lb data sources.

**Comment by Jukie** [comment]
Updated with missing regions and an upstream merge.
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSElbHostedZoneId_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSElbHostedZoneId_basic -timeout 120m
=== RUN   TestAccAWSElbHostedZoneId_basic
=== PAUSE TestAccAWSElbHostedZoneId_basic
=== CONT  TestAccAWSElbHostedZoneId_basic
--- PASS: TestAccAWSElbHostedZoneId_basic (38.50s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	40.879s
```

**Comment by Jukie** [comment]
Any new thoughts on this? Can rework if necessary.

**Comment by ewbankkit** [comment]
Superseded by #24749.
@Jukie Thanks for the contribution :tada: :clap:.

**Pull request closed without merging**